Add support for plotting multiple state files at once

PeterLombaers opened the pull request:
This pull request implement plotting multiple curves in one plot. It consists of three parts:

1. The existing plotting functions have been split into functions that add a curve to a plot, and functions that add information and style the axis. This seems generally useful, because also plot styling can be more easily customized this way.
2. For the three existing plot types there is a 'multiple' version that plot several curves of that type in the same plot.
3. The multiple plots are connected to the command line. Note that the multiple plots take filepaths as input, while the single plots take state objects as input. For the multiple plots it seems undesirable to load all the state objects at the same time, as this could get memory heavy. However, it is a bit ugly that they take different kinds of inputs, so this could be improved.

J535D165 commented:
Thanks for this PR! 

I think there are some things regarding API design and maintainability we can improve. I'm drafting (without testing) an alternative implementation:

## This PR

```python
def plot_recall(ax,
                state_obj,
                priors=False,
                x_absolute=False,
                y_absolute=False,
                show_random=True):
    """Plot the recall@T for all thresholds T.
    Arguments
    ---------
    state_obj: asreview.state.SQLiteState
        State object from which to get the labels for the plot.
    priors: bool
        Include the prior in plot or not.
    x_absolute: bool
        If True, the number of records is on the x-axis.
        If False, the fraction of the whole dataset is on the x-axis.
    y_absolute: bool
        If True, the number of included records found is on the y-axis.
        If False, the fraction of all included records found is on the y-axis.
    show_random: bool
        Show the random curve in the plot.
    Returns
    -------
    matplotlib.axes.Axes
    Notes
    -----
    The recall at T statistic is defined as the number of relevant records
    found after reviewing T records.
    """

    labels = get_labels(state_obj)

    return _plot_recall(ax,
                        labels,
                        x_absolute=x_absolute,
                        y_absolute=y_absolute)


def plot_multiple_recall(ax,
                         project_fps,
                         priors=False,
                         x_absolute=False,
                         y_absolute=False,
                         show_random=True,
                         show_legend=True):
    """Plot the recall@T for all thresholds T, for multiple state objects.
    Arguments
    ---------
    project_fps: list[Path]
        List of filepaths of ASReview project files.
    priors: bool
        Include the prior in plot or not.
    x_absolute: bool
        If True, the number of records is on the x-axis.
        If False, the fraction of the whole dataset is on the x-axis.
    y_absolute: bool
        If True, the number of included records found is on the y-axis.
        If False, the fraction of all included records found is on the y-axis.
    show_random: bool
        Show the random curve in the plot.
    show_legend: bool
        Show the legend in the plot.
    Returns
    -------
    matplotlib.axes.Axes
    """
    labels_dict = get_multiple_labels(project_fps)

    return _plot_multiple_recall(ax,
                                 labels_dict,
                                 x_absolute,
                                 y_absolute,
                                 show_random,
                                 show_legend)
```

## Suggestion

```python

def plot_recall(ax,
                         state_obj,
                         priors=False,
                         x_absolute=False,
                         y_absolute=False,
                         show_random=True,
                         legend_values=None,
                         show_legend=True):
    """Plot the recall@T for all thresholds T, for multiple state objects.
    Arguments
    ---------
    state_obj: list[Path]
        List of filepaths of ASReview project files.
    priors: bool
        Include the prior in plot or not.
    x_absolute: bool
        If True, the number of records is on the x-axis.
        If False, the fraction of the whole dataset is on the x-axis.
    y_absolute: bool
        If True, the number of included records found is on the y-axis.
        If False, the fraction of all included records found is on the y-axis.
    show_random: bool
        Show the random curve in the plot.
    show_legend: bool
        Show the legend in the plot.
    Returns
    -------
    matplotlib.axes.Axes
    """
    labels = get_labels(state_obj)  # return list of lists when multiple state objects are given!! (no need to return generator here)

    return _plot_multiple_recall(ax,
                                 labels,
                                 x_absolute,
                                 y_absolute,
                                 show_random,
                                 legend_values=legend_values,
                                 show_legend)
```

## This PR

```python
# Plotting using labels.
def _plot_recall(ax,
                 labels,
                 x_absolute=False,
                 y_absolute=False,
                 show_random=True):
    """Plot the recall of a state object, using the labels.
    labels: list
        List of labels.
    """
    ax = _add_recall_curve(ax, labels, x_absolute, y_absolute)
    ax = _add_recall_info(ax, labels, x_absolute, y_absolute)
    if show_random:
        ax = _add_random_curve(ax, labels, x_absolute, y_absolute)

    return ax


def _plot_multiple_recall(ax,
                          labels_dict,
                          x_absolute=False,
                          y_absolute=False,
                          show_random=True,
                          show_legend=True,
                          legend_kwargs={}):
    """Plot multiple recall curves, using the labels.
    labels_dict : dict[str, list]
        Dictionary {state_name: list_of_labels}
    """
    for legend_label, labels in labels_dict.items():
        ax = _add_recall_curve(ax, labels, x_absolute, y_absolute, legend_label)
    ax = _add_recall_info(ax, labels, x_absolute, y_absolute)

    if show_random:
        ax = _add_random_curve(ax, labels, x_absolute, y_absolute)

    if show_legend:
        ax.legend(**legend_kwargs)

    return ax

```

## Suggestion

```python

def _plot_recall(ax,
                          labels,
                          x_absolute=False,
                          y_absolute=False,
                          show_random=True,
                          show_legend=True,
                          legend_values=None,
                          legend_kwargs={}):
    """Plot multiple recall curves, using the labels.
    labels_dict : list[list]
        Dictionary {state_name: list_of_labels}
    """
    
    if labels is not LIST OF LISTS:
        labels= [labels]
        show_legend=False
    
    for i,label_set in enumerate(labels):
        ax = _add_recall_curve(ax, label_set, x_absolute, y_absolute, legend_values[i])
    ax = _add_recall_info(ax, label_set, x_absolute, y_absolute)

    if show_random:
        ax = _add_random_curve(ax, label_set, x_absolute, y_absolute)

    if show_legend:
        ax.legend(**legend_kwargs)

    return ax

```

Why this approach? Less duplicated code and docstrings. API and CLI are more in line with each other. 

## Issue 3

By using lists instead of dicts, we are not only closer to the matplotlib API, but we can also use generators instead of lists. For example:

```python
def iter_states(file_paths):
    with open_state(file_paths) as s:
        yield s
```

PeterLombaers commented:
Nice suggestions. I especially like using a generator. That makes dealing with a single state and multiple states a lot easier.

J535D165 commented on asreviewcontrib/insights/plot.py:
```diff
@@ -52,85 +53,27 @@ def plot_recall(ax,
     ---------
-    state_obj: asreview.state.SQLiteState
-        State object from which to get the labels for the plot.
+    state_obj: Union[asreview.state.SQLiteState,
+                     list[asreview.state.SQLiteState]]
```
I think documenting the types is not so common. I think sphinx can handle typing nowadays very well. You can skip this if you apply typing. I would suggest using traditional docstrings and doing proper typing in another PR.

J535D165 commented on asreviewcontrib/insights/utils.py:
```diff
@@ -1,8 +1,33 @@
+        List of labels if state_obj is a single state. List of lists of labels
+        if state_obj is a list of states.
+    """
+    print(type(state_obj))
```
```suggestion
```

J535D165 commented on asreviewcontrib/insights/utils.py:
```diff
@@ -1,8 +1,33 @@
+        if state_obj is a list of states.
+    """
+    print(type(state_obj))
+    if isinstance(state_obj, SQLiteState):
+        return _get_labels(state_obj, priors)
+    else:
+        return [_get_labels(single_state, priors) for single_state in state_obj]
+
```
I prefer recursive functions 

```suggestion
    if isinstance(state_obj, SQLiteState):
        n_records = state_obj.n_records
    
        # get the labels
        labels = state_obj.get_labels(priors=priors).to_list()
        # if less labels than records, check if all labels available
        if len(labels) < n_records:
            labels = labels + np.zeros(n_records - len(labels)).tolist()
    
        return labels
    else:
        return [get_labels(single_state, priors) for single_state in state_obj]

```

Something like this

J535D165 commented on asreviewcontrib/insights/utils.py:
```diff
@@ -1,8 +1,33 @@
+from asreview.state import SQLiteState
 
 
 def get_labels(state_obj, priors=False):
```
Maybe better to keep this function private. To avoid users using this module for this function in particular.

J535D165 commented on asreviewcontrib/insights/utils.py:
```diff
@@ -15,3 +40,21 @@ def get_labels(state_obj, priors=False):
     return labels
+
+
+def iter_states(file_paths):
```
Also private?

PeterLombaers replied:
I noticed when making the API example for multiple curves with a custom legend, that I wanted to use this function, but it was private. Maybe there's an argument for keeping this public, though I agree that you don't want people to use this function outside of this specific case. So I made it private in the update.

J535D165 reviewed (changes requested):
Thanks for the update!

PeterLombaers commented:
I incorporated the suggestions and updated the README as well.

JTeijema commented:
This looks good! I do think that [asreview/asreview-makita/pull/4](https://github.com/asreview/asreview-makita/pull/4) will break if this is merged. The fix shouldn't be complicated though.

PeterLombaers commented:
I've updated the pull request with the latest changes from master. I also added support for multiple files in the metrics endpoint. I think this is ready to be merged.

J535D165 commented:
Very nice work
